x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/role/RoleWithRemoteIndicesPrivilegesRestIT.java

```diff
@@ -229,7 +255,65 @@ public void testGetUserPrivileges() throws IOException {
                   "clusters": ["remote-a", "*"]
                 }
               ]
-            }""", false)));
+            }""")));
+    }
+
+    public void testGetUserPrivilegesWithMultipleFlsDlsDefinitionsPreservesGroupPerIndexPrivilege() throws IOException {
+        final var putRoleRequest = new Request("PUT", "/_security/role/" + REMOTE_SEARCH_ROLE);
+        putRoleRequest.setJsonEntity("""
+            {
+              "remote_indices": [
+                {
+                  "names": ["index-a", "*"],
+                  "privileges": ["read"],
+                  "clusters": ["remote-a", "*"],
+                  "query": "{\\"match\\":{\\"field-a\\":\\"a\\"}}",
+                  "field_security": {
+                    "grant": ["field-a"]
+                  }
+                },
+                {
+                  "names": ["index-a", "*"],
+                  "privileges": ["read"],
+                  "clusters": ["remote-a", "*"],
+                  "query": "{\\"match\\":{\\"field-b\\":\\"b\\"}}",
+                  "field_security": {
+                    "grant": ["field-b"]
+                  }
+                }
+              ]
+            }""");
+        final Response putRoleResponse1 = adminClient().performRequest(putRoleRequest);
+        assertOK(putRoleResponse1);
+
+        final Response getUserPrivilegesResponse1 = executeAsRemoteSearchUser(new Request("GET", "/_security/user/_privileges"));
+        assertOK(getUserPrivilegesResponse1);
+        final Map<String, Object> actual = responseAsMap(getUserPrivilegesResponse1);
+        // The order of remote indices is not deterministic, so manually extract remote indices from map response to compare ignoring order
+        @SuppressWarnings("unchecked")
+        final List<Object> rawRemoteIndices = (List<Object>) actual.get("remote_indices");
+        assertThat(rawRemoteIndices, notNullValue());
+        assertThat(rawRemoteIndices, containsInAnyOrder(mapFromJson("""
```

Comment:
I'll simplify this in https://github.com/elastic/elasticsearch/pull/91734, once I've made order deterministic there.